File: go/factory.go
package hctmcpsignals

import "time"

// Option is a functional option for creating signals.
type Option func(*HCTSignal)

// WithUrgency sets the urgency level (1-10).
func WithUrgency(u int) Option {
	return func(s *HCTSignal) {
		if s.Performance == nil {
			s.Performance = &Performance{}
		}
		if u < 1 {
			u = 1
		}
		if u > 10 {
			u = 10
		}
		s.Performance.Urgency = u
	}
}

// WithTempo sets the tempo.
func WithTempo(t Tempo) Option {
	return func(s *HCTSignal) {
		if s.Performance == nil {
			s.Performance = &Performance{}
		}
		s.Performance.Tempo = t
	}
}

// WithTimeoutMs sets the timeout in milliseconds.
func WithTimeoutMs(ms int64) Option {
	return func(s *HCTSignal) {
		if s.Performance == nil {
			s.Performance = &Performance{}
		}
		s.Performance.TimeoutMs = &ms
	}
}

// WithPayload sets the payload.
func WithPayload(p map[string]interface{}) Option {
	return func(s *HCTSignal) {
		s.Payload = p
	}
}

// WithPayloadEntry adds a single payload entry.
func WithPayloadEntry(key string, value interface{}) Option {
	return func(s *HCTSignal) {
		if s.Payload == nil {
			s.Payload = make(map[string]interface{})
		}
		s.Payload[key] = value
	}
}

// WithHoldType sets the hold type for FERMATA signals.
func WithHoldType(ht HoldType) Option {
	return func(s *HCTSignal) {
		if s.Conditions == nil {
			s.Conditions = &Conditions{}
		}
		s.Conditions.HoldType = &ht
	}
}

// WithRepeatUntil sets the repeat condition for VAMP signals.
func WithRepeatUntil(condition string) Option {
	return func(s *HCTSignal) {
		if s.Conditions == nil {
			s.Conditions = &Conditions{}
		}
		s.Conditions.RepeatUntil = &condition
	}
}

// WithQualityThreshold sets the quality threshold (0-1).
func WithQualityThreshold(threshold float64) Option {
	return func(s *HCTSignal) {
		if s.Conditions == nil {
			s.Conditions = &Conditions{}
		}
		s.Conditions.QualityThreshold = &threshold
	}
}

// newSignal creates a base signal.
func newSignal(signalType SignalType, source string, targets []string, opts ...Option) *HCTSignal {
	s := &HCTSignal{
		Type:    signalType,
		Source:  source,
		Targets: targets,
		Payload: make(map[string]interface{}),
		Performance: &Performance{
			Urgency: 5,
			Tempo:   Moderato,
		},
		Timestamp: time.Now().UTC(),
	}
	for _, opt := range opts {
		opt(s)
	}
	return s
}

// NewCue creates a CUE signal to trigger agent activation.
func NewCue(source string, targets []string, opts ...Option) *HCTSignal {
	return newSignal(Cue, source, targets, opts...)
}

// NewFermata creates a FERMATA signal to hold for approval.
func NewFermata(source, reason string, opts ...Option) *HCTSignal {
	allOpts := append([]Option{
		WithPayloadEntry("reason", reason),
		WithHoldType(HoldHuman),
	}, opts...)
	return newSignal(Fermata, source, []string{"governance"}, allOpts...)
}

// NewAttacca creates an ATTACCA signal for immediate transition.
func NewAttacca(source string, targets []string, opts ...Option) *HCTSignal {
	allOpts := append([]Option{
		WithUrgency(10),
		WithTempo(Presto),
	}, opts...)
	return newSignal(Attacca, source, targets, allOpts...)
}

// NewVamp creates a VAMP signal to repeat until condition met.
func NewVamp(source, repeatUntil string, opts ...Option) *HCTSignal {
	threshold := 0.9
	timeout := int64(60000)
	allOpts := append([]Option{
		WithRepeatUntil(repeatUntil),
		WithQualityThreshold(threshold),
		WithTimeoutMs(timeout),
	}, opts...)
	return newSignal(Vamp, source, []string{source}, allOpts...)
}

// NewCaesura creates a CAESURA signal for full stop.
func NewCaesura(source, reason string, opts ...Option) *HCTSignal {
	allOpts := append([]Option{
		WithPayloadEntry("reason", reason),
		WithUrgency(10),
		WithTempo(Presto),
	}, opts...)
	return newSignal(Caesura, source, []string{"*"}, allOpts...)
}

// NewTacet creates a TACET signal to mark agent as inactive.
func NewTacet(source string, opts ...Option) *HCTSignal {
	return newSignal(Tacet, source, nil, opts...)
}

// NewDownbeat creates a DOWNBEAT signal for global synchronization.
func NewDownbeat(source, syncPoint string, opts ...Option) *HCTSignal {
	allOpts := append([]Option{
		WithPayloadEntry("sync_point", syncPoint),
	}, opts...)
	return newSignal(Downbeat, source, []string{"*"}, allOpts...)
}

File: rust/src/factory.rs
//! Factory functions for creating HCT signals with builder pattern.

use crate::signals::{Conditions, HCTSignal, HoldType, Performance, SignalType, Tempo};
use chrono::Utc;
use std::collections::HashMap;

/// Builder for creating HCT signals.
#[derive(Debug, Clone)]
pub struct SignalBuilder {
    signal_type: SignalType,
    source: String,
    targets: Vec<String>,
    payload: HashMap<String, serde_json::Value>,
    performance: Performance,
    conditions: Option<Conditions>,
}

impl SignalBuilder {
    /// Create a new builder.
    fn new(signal_type: SignalType, source: impl Into<String>) -> Self {
        Self {
            signal_type,
            source: source.into(),
            targets: Vec::new(),
            payload: HashMap::new(),
            performance: Performance::default(),
            conditions: None,
        }
    }

    /// Add targets.
    #[must_use]
    pub fn with_targets(mut self, targets: impl IntoIterator<Item = impl Into<String>>) -> Self {
        self.targets = targets.into_iter().map(Into::into).collect();
        self
    }

    /// Add payload.
    #[must_use]
    pub fn with_payload(mut self, payload: HashMap<String, serde_json::Value>) -> Self {
        self.payload = payload;
        self
    }

    /// Add a single payload entry.
    #[must_use]
    pub fn with_payload_entry(mut self, key: impl Into<String>, value: serde_json::Value) -> Self {
        self.payload.insert(key.into(), value);
        self
    }

    /// Set urgency (1-10).
    #[must_use]
    pub fn with_urgency(mut self, urgency: u8) -> Self {
        self.performance.urgency = urgency.clamp(1, 10);
        self
    }

    /// Set tempo.
    #[must_use]
    pub fn with_tempo(mut self, tempo: Tempo) -> Self {
        self.performance.tempo = tempo;
        self
    }

    /// Set timeout.
    #[must_use]
    pub fn with_timeout_ms(mut self, timeout_ms: u64) -> Self {
        self.performance.timeout_ms = Some(timeout_ms);
        self
    }

    /// Set conditions.
    #[must_use]
    pub fn with_conditions(mut self, conditions: Conditions) -> Self {
        self.conditions = Some(conditions);
        self
    }

    /// Build the signal.
    #[must_use]
    pub fn build(self) -> HCTSignal {
        HCTSignal {
            signal_type: self.signal_type,
            source: self.source,
            targets: self.targets,
            payload: self.payload,
            performance: Some(self.performance),
            conditions: self.conditions,
            timestamp: Utc::now(),
        }
    }
}

/// Create a CUE signal to trigger agent activation.
#[must_use]
pub fn cue(source: impl Into<String>, targets: impl IntoIterator<Item = impl Into<String>>) -> SignalBuilder {
    SignalBuilder::new(SignalType::Cue, source)
        .with_targets(targets)
}

/// Create a FERMATA signal to hold for approval.
#[must_use]
pub fn fermata(source: impl Into<String>, reason: impl Into<String>) -> SignalBuilder {
    SignalBuilder::new(SignalType::Fermata, source)
        .with_targets(["governance"])
        .with_payload_entry("reason", serde_json::Value::String(reason.into()))
        .with_conditions(Conditions {
            hold_type: Some(HoldType::Human),
            ..Default::default()
        })
}

/// Create an ATTACCA signal for immediate transition.
#[must_use]
pub fn attacca(source: impl Into<String>, targets: impl IntoIterator<Item = impl Into<String>>) -> SignalBuilder {
    SignalBuilder::new(SignalType::Attacca, source)
        .with_targets(targets)
        .with_urgency(10)
        .with_tempo(Tempo::Presto)
}

/// Create a VAMP signal to repeat until condition met.
#[must_use]
pub fn vamp(source: impl Into<String>, repeat_until: impl Into<String>) -> SignalBuilder {
    let source_str = source.into();
    SignalBuilder::new(SignalType::Vamp, source_str.clone())
        .with_targets([source_str])
        .with_conditions(Conditions {
            repeat_until: Some(repeat_until.into()),
            quality_threshold: Some(0.9),
            ..Default::default()
        })
        .with_timeout_ms(60_000)
}

/// Create a CAESURA signal for full stop.
#[must_use]
pub fn caesura(source: impl Into<String>, reason: impl Into<String>) -> SignalBuilder {
    SignalBuilder::new(SignalType::Caesura, source)
        .with_targets(["*"])
        .with_payload_entry("reason", serde_json::Value::String(reason.into()))
        .with_urgency(10)
        .with_tempo(Tempo::Presto)
}

/// Create a TACET signal to mark agent as inactive.
#[must_use]
pub fn tacet(source: impl Into<String>) -> SignalBuilder {
    SignalBuilder::new(SignalType::Tacet, source)
}

/// Create a DOWNBEAT signal for global synchronization.
#[must_use]
pub fn downbeat(source: impl Into<String>, sync_point: impl Into<String>) -> SignalBuilder {
    SignalBuilder::new(SignalType::Downbeat, source)
        .with_targets(["*"])
        .with_payload_entry("sync_point", serde_json::Value::String(sync_point.into()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cue_builder() {
        let signal = cue("orchestrator", ["analyst", "synthesizer"])
            .with_urgency(8)
            .with_tempo(Tempo::Allegro)
            .with_payload_entry("task", serde_json::json!("Analyze Q4"))
            .build();

        assert_eq!(signal.signal_type, SignalType::Cue);
        assert_eq!(signal.source, "orchestrator");
        assert_eq!(signal.targets, vec!["analyst", "synthesizer"]);
        assert_eq!(signal.performance.as_ref().map(|p| p.urgency), Some(8));
    }

    #[test]
    fn test_fermata_builder() {
        let signal = fermata("reporter", "Ready for review").build();

        assert_eq!(signal.signal_type, SignalType::Fermata);
        assert!(signal.payload.contains_key("reason"));
        assert!(signal.conditions.is_some());
    }

    #[test]
    fn test_attacca_is_urgent() {
        let signal = attacca("agent", ["next"]).build();

        assert_eq!(signal.performance.as_ref().map(|p| p.urgency), Some(10));
        assert_eq!(signal.performance.as_ref().map(|p| p.tempo), Some(Tempo::Presto));
    }

    #[test]
    fn test_caesura_broadcasts() {
        let signal = caesura("governance", "Budget exceeded").build();

        assert_eq!(signal.targets, vec!["*"]);
    }
}
